Usability improvements

charliebruce opened the pull request:
Add a bunch of minor usability tweaks:

* Defaults to .brd file type on Linux, OS X and Windows, with an option for "All Files" if your board file has an odd extension (untested on anything except Windows)
* Shows error messages if DirectX fails to initialise
* Adds keyboard controls to move the viewport around (caveat: behaves oddly when rotated at 90 and 270 degrees - will refactor/fix this soon)
* Drag and drop to open file on Windows
* Window title shows filename on Windows

inflex commented:
Already have the keyboard pan controls sorted out to deal with the 90/270/flip problem.

I'm just waiting for a stable core base before I start rolling all my fixes back in to it.

MattSturgeon commented:
@charliebruce Thanks for taking the time to refactor your changes, and adding file filters to the other `show_file_picker` implementations! I'll get around to looking at this within a few hours.

@inflex were your changes based on @charliebruce's original PRs? If so we should be able to get this in and then add your patches on top with the GLFW port.

It'd be great if someone with OSX access could test 61ec392fe885588684affb09e991adc04ff1ca50 on a branch that builds for mac. (@piernov ?)

piernov commented:
61ec392fe885588684affb09e991adc04ff1ca50 seems to work on OS X after cherry-picking it into my repo.

Since we support other file format it would be nice to put the file extensions in a vector (or even a map for giving a reference to the function handling it) and loop through it. No need to hardcode them in 4 different places.

MattSturgeon commented:
> 61ec392 seems to work on OS X after cherry-picking it into my repo.

Good stuff

> Since we support other file format it would be nice to put the file
extensions in a vector

I've seen it done like so:

(Will edit in from pc)

On Wed, 6 Jul 2016, 10:38 piernov, <notifications@github.com> wrote:

> 61ec392
> <https://github.com/OpenBoardView/OpenBoardView/commit/61ec392fe885588684affb09e991adc04ff1ca50>
> seems to work on OS X after cherry-picking it into my repo.
>
> Since we support other file format it would be nice to put the file
> extensions in a vector (or even a map for giving a reference to the
> function handling it) and loop through it. No need to hardcode them in 4
> different places.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenBoardView/OpenBoardView/pull/25#issuecomment-230725354>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE0BItALDK8QZbcjIPNvfSKPzwsIdVTGks5qS3eRgaJpZM4JFr4D>
> .
>

MattSturgeon commented:
@charliebruce 

#### Feedback on the keyboard shortcuts:

 - For me personally the "default" jump distance is a bit too far, I find myself uncomfortably holding `ctrl` all the time! I's suggest you could decrease the default movement to be closer to slow-mode, and add a fast-mode toggled by `shift`.
 - The zoom in/out has no speed adjustment. Keyboard/mouse zoom speed should work the same as panning around.
   - `I`/`O` will probably need changing as `Ctrl+O` maps to open.
   - Numpad `+`/`-` should work the same as normal `+`/`-`
 - Would be good to add `Ctrl+Q` for quitting

Otherwise nice work!

#### Other bits

 - Drag 'n' Drop: Works well!
    - I'd prefer if the "multiple files" error message was consistent with other dialogues like the "Error opening file" one (see below)
 - Could you abstract `MessageBox` into a `show_error_msg` wrapper in the platform files as this looks windows specific.
   - Also would be good to move the other errors (I can think of "error loading file") to use this new function.

 - I was unable to test the d3d error messages, but I assume it's the same as the one reported working in #22. I don't know yet how much of this will be useful once we've dumped dx9 for GL, but it's affecting users now so

> keyboard controls to move the viewport around (caveat: behaves oddly when rotated at 90 and 270 degrees - will refactor/fix this soon)

It would be great if you got this sorted, I believe @inflex had done some work on this that could help (albeit based off an early linux port)

-----------------------

I quite like this on the whole, the `BoardView::ChangeZoom` and `BoardView::OpenFile` abstractions are nice too.

Final notes for now, it'd be good if you rebased out some of the extra commits from the history (like the merge commits). Also I haven't checked, but have you run [`clang-format.py`](https://llvm.org/svn/llvm-project/cfe/trunk/tools/clang-format/) on your changes yet?

-------------------------------
@piernov 
> Since we support other file formats it would be nice to put the file extensions in an std::vector (or even an std::map for giving a reference to the method handling it) and loop through it. No need to hardcode them in 4 different places.

Misread this earlier. Yes, it would be good to move to a unified filter function. I'm not sure this is needed for the initial implementation though.

pafmaf commented:
> `[op setAllowedFileTypes:@[@"brd", @"other", @"file", @"types"]];`

As I understood it

> for giving a _reference_ to the _method_ handling it and loop through it

means having something like a factory pattern with `Board`s, i.e. different file types have to register somewhere and the factory _method_ instantiates the right thing. `Board::getRegisteredFileTypes()`

MattSturgeon commented:
@pafmaf @piernov 

> As I understood it
> 
>     for giving a reference to the method handling it and loop through it
> 
> means having something like a factory pattern with Boards, i.e. different file types have to register somewhere and the factory method instantiates the right thing. Board::getRegisteredFileTypes()

Yes I missed that he was talking about a map/filter functions - doh! (see the note at the end of my last [comment](#issuecomment-230737865) - added it just as you posted!)

charliebruce commented:
Have made some changes:
* Have added Ctrl-Q as a quit shortcut.
* Refactored error-handling.
* Added numpad controls, also PgUp/PgDn zoom in/out.
* Have halved default movement speed, with Ctrl working as before (decreasing movement speed by factor of 10). Feels easier to use now. I personally prefer Ctrl-to-slow-down rather than Shift-to-speed-up, is there a standard/expected behaviour in the industry or is this up to preference?

I haven't fully addressed all points though:
* Haven't added speed control on keyboard zooming - that level of precision when zooming doesn't feel necessary to me, and allows the idiom I/O for In/Out to be kept, without breaking the Ctrl-O shortcut. As before, if there's an expected behaviour/use case for it can easily add though.
* Haven't tested numpad yet - my primary machine is a laptop so I don't have one to test with unless I dig up an external USB one.
* Have kept the Direct3D errors in MessageBoxes because they occur before ImGui has loaded so can't rely on it to show errors, and will necessarily be Windows-specific anyway.
* Probably not worth refactoring file type filtering for now, can sort that once we have a larger set of supported file types/importers.

Will do over weekend:
* 90/270 rotation fix.
* Don't have Python set up on this machine, haven't run clang-format.py yet.
* Rebase/squash to tidy everything up.

MattSturgeon commented on BoardView.cpp:
```diff
@@ -334,38 +334,58 @@ void BoardView::HandleInput() {
-		// (I/+) and (O/-) keys as alternate zoom. Will zoom on the centre of the viewport, not on the mouse.
-		if (ImGui::IsKeyPressed('I') || ImGui::IsKeyPressed(187)) {
+		//Ctrl-Q as alternative quit shortcut (Alt-F4 is already handled by Windows)
+		if (ImGui::IsKeyPressed('Q') && ImGui::IsKeyDown(17)) {
```
Could you move this up to line 41 so it is together with `Ctrl+O`?

MattSturgeon commented on BoardView.cpp:
```diff
@@ -334,38 +334,58 @@ void BoardView::HandleInput() {
+		
+		//TODO numpad +/- scancodes
+		if ( ImGui::IsKeyPressed('I')
+			|| ImGui::IsKeyPressed(187)
```
If possible, could you replace all hard coded numbers with `GetKeyIndex`?  That should aid readability (and porting to GLFW later)

MattSturgeon commented on BoardView.cpp:
```diff
@@ -28,6 +28,11 @@ BoardView::~BoardView() {
 }
 
+void BoardView::ShowError(char* msg) {
+	m_lastErrorMsg = msg;
```
Great, this refrac should make it easier to drop `ImGui::OpenPopup` for errors and just use native dialogs for all errors.

Would mean wrapping `MessageBox`, and Gtk + Cocoa equivalents in a cross-platform `show_error` function though, so happy to leave it for now.

MattSturgeon commented on platform_unix.cpp:
```diff
@@ -51,6 +51,11 @@ char *show_file_picker() {
+	// Filter file types
+	GtkFileFilter *filter = gtk_file_filter_new();
+	gtk_file_filter_add_pattern(filter, "*.brd");
+	gtk_file_chooser_add_filter(dialog, filter);
```
This needs to be `gtk_file_chooser_add_filter(GTK_FILE_CHOOSER(dialog), filter);` - turns out the dialog needs casting to the `GtkFileChooser` interface to use `gtk_file_chooser_add_filter` on it.

MattSturgeon replied:
For feature parity with windows, you should also use `gtk_file_filter_set_name(filter, "BRD Files");` and consider adding a second GtkFileFilter with pattern `"*"` and name `"All Files"`

MattSturgeon commented on BoardView.cpp:
```diff
@@ -320,22 +326,87 @@ void BoardView::HandleInput() {
+		//Ctrl-Q as alternative quit shortcut (Alt-F4 is already handled by Windows)
+		if (ImGui::IsKeyPressed('Q') && ImGui::IsKeyDown(17)) {
+			m_wantsQuit = true;
+		}
```
Could you have `Ctrl+O` and `Ctrl+Q` together? Would probably mean moving this to the top of  `Update` method.

MattSturgeon commented on platform_osx.mm:
```diff
@@ -7,6 +7,10 @@
 
+	// Filter file types
+	[op setCanChooseFiles:YES];
+	[op setAllowedFileTypes:@[@"brd"]];
```
May be worth adding `[op setAllowsMultipleSelection:FALSE];`

MattSturgeon commented:
Hi @charliebruce, appologies if you've been put off by some of the feedback (or the volume of it). I'd still like to get this in, are you still planning to work on this?

If not I am happy to make the final adjustments needed to get these changes merged.

charliebruce commented:
Hi @MattSturgeon - no problems with the feedback, have just been quite busy with other projects over the last couple of weeks. I can take another look at all this stuff on Friday/Saturday evening, unless you have any free time before then and want to get it done sooner - in that case, feel free to build on my changes or cherry-pick anything that is useful to you.

MattSturgeon commented:
@charliebruce I've split out the keyboard stuff from the rest in [charlie-part1] and [charlie-part2]. I've invited you to a new "Charliebruce PR" team to give you push access to the new branches.

Feel free to work off of those branches and replace this PR.

I've not tested these yet, so hopefully I haven't messed up anything with my rebasing. (I'll edit this line once I've tested the new branches)

[charlie-part1]: https://github.com/OpenBoardView/OpenBoardView/commits/charlie-part1
[charlie-part2]: https://github.com/OpenBoardView/OpenBoardView/commits/charlie-part2

MattSturgeon commented:
@charliebruce I've merged part1 and added some changes to part2.

### [Changes to part2](https://github.com/OpenBoardView/OpenBoardView/compare/b42316e...charlie-part2)

- Moved `Ctrl+O` from `Update()` into `HandleInput()`
  - The `open_file` variable is now the property `m_open_file`
- Added a generic speed modifier
  - `Ctrl` will divide speed by 5 (`speed = 0.2`)
  - `Shift` will multiply speed by 5 (`speed = 5`)
  - Tweaked default `distance`, `zoom` and `mwheel` and applied `speed` to each of them.
- Minor changes
  - Fixed brackets around checking `GetKeyIndex` (fixes page up/down)
  - Use `io.KeyCtrl` instead of `IsKeyDown(17)`
  - moved `powof` into `ChangeZoom()`: positive `factor` will zoom in, negative out. This makes it easier to apply `speed` modifier.

See the [diff](https://github.com/OpenBoardView/OpenBoardView/compare/b42316e...charlie-part2) for full changes

### [Merged into master](https://github.com/OpenBoardView/OpenBoardView/commits/master)

- Refactor error handling
- Display an error when DirectX can't be initialised
  - Rather than silently exiting, we should inform the user if DirectX can't
     be initialised.
- Add drag and drop
- Display filename in window title
- Add .brd filetype filter to the Open File dialog
  - On Windows and Gtk (linux) we add a drop down to choose between any file
     or just .brd files. On OSX we tell the dialog that we only want .brd
     files.

MattSturgeon commented:
Merged the last part as 4d245d94be0807c9402a207ad1cf8e7d57f5fd95. In addition to the changes noted above, I also changed a few comments and reordered the contents of `HandleInput()` to group all shortcuts together and all controls together.
